x-pack/platform/plugins/shared/fleet/server/services/epm/packages/update_custom_integration.ts

```
@@ -0,0 +1,188 @@
+
+// Define a type for the integration attributes
+interface IntegrationAttributes {
+  version: string;
+  install_source: string;
+  [key: string]: any;
+}
+
+import { getInstalledPackageWithAssets } from './get';
+
+import { installPackageWithStateMachine } from './install';
+import { CustomIntegrationNotFoundError } from './custom_integrations/validation/check_custom_integration';
+
+export async function updateCustomIntegration(
+  esClient: ElasticsearchClient,
+  soClient: SavedObjectsClientContract,
+  id: string,
+  fields: {
+    readMeData?: string;
+    categories?: string[];
+  }
+) {
+  try {
+    // Get the current integration using the id
+    const integration: SavedObject<IntegrationAttributes> = await soClient.get(
+      PACKAGES_SAVED_OBJECT_TYPE,
+      id
+    );
+    if (!integration) {
+      throw new Error(`Integration with ID ${id} not found`);
+    } else if (
+      integration.attributes.install_source !== 'custom' &&
+      integration.attributes.install_source !== 'upload'
+    ) {
+      throw new CustomIntegrationNotFoundError(
+        `Integration with ID ${id} is not a custom integration`
+      );
+    } else {
+      // add one to the patch version in the semver
+      const newVersion = integration.attributes.version.split('.');
+      newVersion[2] = (parseInt(newVersion[2], 10) + 1).toString();
+      const newVersionString = newVersion.join('.');
+      // Increment the version of everything and create a new package
+      const res = await incrementVersionAndUpdate(soClient, esClient, id, {
+        version: newVersionString,
+        readme: fields.readMeData,
+      })
+        .then((response) => {
+          return response;
+        })
+        .catch((e) => {
+          return e;
+        });
+      return {
+        version: newVersionString,
+        status: res.status,
+      };
+    }
+  } catch (error) {
+    throw new Error(error.message);
```

Not sure that catch is needed no?

---

No, it will be caught at the handler level, no need to catch and just throw again. Good catch!

---

I removed that catch, but in order to catch any errors thrown by the SavedObject client, I wrapped it in a catch and ignored the error so that we could then return a custom error when an integration doesn't exist, rather than the error thrown by the SO client.